RFC3164

leodido opened the pull request:
This PR finally implements RFC3164 parsing! 😃 

It fixes #15 when merged in.

Notice it changes the `syslog.Message` interface and it introduces a `syslog.Base` struct containing the common fields between the two Syslog formats.

This means I'll need to release another major (v3) when completed 100% (see below).

Missing pieces at the moment:

- [ ] extensive test suite
- [x] option/mechanism to let the user choose the year of the timestamp
- [x] best-effort mode
- [x] option to choose the year for STAMP timestamps (`WithYear()` with strategies like `CurrentYear{}`, `Year{YYYY:2020}` etc.)
- [x] option to choose the time zone for STAMP timestamps (`WithTimezone()`)
- [x] option to let the parser ALSO match RFC3339 timestamps (`WithRFC3339()`)
- [x] examples
- [x] update README with examples referring to the new interfaces (#26)

leodido committed:
Build: RFC3164 rule

leodido committed:
New: RFC3164 initial impl.

leodido committed:
Fix: Priority value for RFC3164

leodido committed:
Fix: Porting initial RFC3164 FSM to new interface

leodido committed:
New: Tag and content message part for RFC3164

leodido committed:
New: Parser structure for RFC3164

leodido committed:
Build: Makefile for RFC3164

leodido committed:
Update: RFC5424

leodido committed:
Fix: Typos and types glitches in RFC3164

leodido committed:
New: RFC3164

leodido committed:
New: Best effort option for RFC3164

leodido committed:
Chore: Syslog message private struct (RFC3164)

leodido committed:
New: RFC3164 FSM tests setup

leodido committed:
Update: Functions to check priority and version ranges

leodido committed:
Update: RFC5424 parser and builer to use common functions for priority and version ranges

leodido committed:
Update: RFC3164 syslog message

gianarb commented:
Hello @leodido !
Thank you for this contrib

```
--- FAIL: TestMachineParse (0.00s)
    --- FAIL: TestMachineParse/<34>Jan_12_06:30:00_xxx_apache_server:_1.2.3.4_-_- (0.00s)
        <autogenerated>:1: 
            	Error Trace:	machine_test.go:49
            	Error:      	Should NOT be empty, but was &{<nil> <nil> <nil> 0 <nil> <nil> <nil> <nil> <nil> <nil> <nil>}
            	Test:       	TestMachineParse/<34>Jan_12_06:30:00_xxx_apache_server:_1.2.3.4_-_-
FAIL
```
This is the failing test atm

leodido commented:
This was the reason I put the PR as a draft (with WIP prefix). Anyway I have other commits to push :)

leodido committed:
update(common): moving out facilities

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
update(common): moving out severities

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
new: introducing base struct containing common fields between syslog formats

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
update(rfc5424): update RFC5424 syslog message to use the base one

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
update(rfc5424): update the RFC5424 builder to the new  RFC5424 syslog messag

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
update(rfc5424): update builder tests

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
update(rfc5424): update test suite to the new API

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
docs(rfc5424): update examples to the new rfc5424 syslog message struct

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
docs(octetcounting): update examples to the new struct

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
docs(nontransparent): update examples to the new syslog message structure

leodido committed:
fix(rfc3164): make tag part less restrictive

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
update(rfc3164): tests

leodido committed:
update(rfc3164): RFC3164 syslog message

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido commented:
greeeeeeeny 💚

leodido committed:
Build: generate dot files for RFC3164

leodido committed:
Docs: RFC3164 *.dot files

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido commented:
@goller I want your review my friend plzzz 🤗 

But please DO NOT merge yet :)

goller commented:
@leodido you rock!

I think I'll start by reading through the structural changes.

Next, I'll read 3164 again and look closely at the grammar.

goller commented on common/facility.go:
```diff
@@ -0,0 +1,53 @@
+	12: "NTP subsystem",
+	13: "log audit",
+	14: "log alert",
+	15: "clock daemon (note 2)", // (todo) > some sources reporting "scheduling daemon"
```
Should this (note 2) be removed?  The others with notes in the RFC that have notes are not specified here.

leodido replied:
I want to remove it too and I'll remove it.

Anyway I left `note 2` here in the past because there are two keys (9, 15) pointing to 2 clock daemons, as you also noted below.

leodido replied:
Done

goller commented on common/facility.go:
```diff
@@ -0,0 +1,53 @@
+	5:  "syslog",
+	6:  "lpr",
+	7:  "news",
+	8:  "uucp",
```
I'm seeing cron also defined as 9 in openbsd:
https://github.com/openbsd/src/blob/master/sys/sys/syslog.h#L98

leodido replied:
Yup, I confirm thanks for noticing it Chris :)

leodido replied:
Done, please take another look at it :)

goller commented on common/facility.go:
```diff
@@ -0,0 +1,53 @@
+	7:  "news",
+	8:  "uucp",
+	10: "authpriv",
+	11: "ftp",
```
[FreeBSD](https://github.com/freebsd/freebsd/blob/master/sys/sys/syslog.h#L107-L109) has 12, 13, and 14 defined as `ntp`, `security`, `console`, respectively.

Think those should be added?

leodido replied:
Yes, and [RFC 5427](https://tools.ietf.org/html/rfc5427) - "Textual Conventions for Syslog Management" - too.

Thus I am going to use this RFC 5427 (fallbacking to RFC 5424 when needed).

leodido replied:
Should be 👌 now

goller commented on common/functions.go:
```diff
@@ -40,3 +40,18 @@ func EscapeBytes(value string) string {
+
+// ValidPriority checks whether the given value is in the priority range [0, 191].
+func ValidPriority(priority uint8) bool {
+	return InBetween(int(priority), 0, 191)
```
Is this needed?  The grammar should forbid any numbers outside of this range.

Is it used for validation of a created and not parsed `SyslogMessage` struct?

leodido replied:
This is needed in 2 places:

- RFC 5424 builder
- RFC 5424 best-effort parsing mode uses it to understand whether the message is partial but still minimally correct or to reject completely (also in best-effort mode). This because such parsing option is technically an error that gets recovered if possible, except when the message was not even minimally indeed :)

goller commented on common/common.rl:
```diff
@@ -16,14 +16,19 @@ bs = 0x5C;
 
 # 01..31
-datemday = ('0' . '1'..'9' | '1'..'2' . '0'..'9' | '3' . '0'..'1');
+datemday = ('0' . nonzerodigit | '1'..'2' . '0'..'9' | '3' . '0'..'1');
```
I think the grammar may need to change.

The RFC says:

```
         If the day of the month is less
         than 10, then it MUST be represented as a space and then the
         number.  For example, the 7th day of August would be
         represented as "Aug  7", with two spaces between the "g" and
         the "7".

```

so, I think it should be

```
 datemday = (sp . nonzerodigit | '1'..'2' . '0'..'9' | '3' . '0'..'1');
```

Is that right?

leodido replied:
It is!

Good catch Chris, fixed :)

goller commented on rfc3164/machine.go.rl:
```diff
@@ -0,0 +1,204 @@
+}
+
+action set_timestamp {
+    if t, e := time.Parse(time.Stamp, string(m.text())); e != nil {
```
The time.Time would be on year 0 with this type of parse:

https://play.golang.org/p/A6ueWMWKPE9

To get the time correctly, one would need to add the correct year.

In reading [Section 4.2](https://tools.ietf.org/html/rfc3164#section-4.2) I see:

```
         If the originally formed message has a TIMESTAMP in the HEADER
         part, then it SHOULD be the local time of the device within its
         timezone.
```

I don't know how one would know the device's timezone.  Perhaps the parser could have a [Location](https://golang.org/pkg/time/#Location) parameter?

leodido replied:
Yeah this was something I had to do but then I forgot. Thanks for reminding me.

My idea was to create an option for this parser - eg., `WithYear()` so that the user can decide the behavior.

WDYT?

goller replied:
I looked around at logstash and there have been a few attempts to make heuristics.
https://github.com/logstash-plugins/logstash-filter-date/pull/4#issue-26908778

Here is another heuristic:
https://github.com/logstash-plugins/logstash-filter-date/issues/51#issuecomment-309866094

This second link has three configurable heuristic strategies.  With your idea of `WithYear()` would be similar to `disambiguate=from_default`

I personally would use `nearest_default` to handle data coming in between Dec and Jan.

leodido replied:
These strategies are cool.

I just pushed a strategy pattern for year strategies and implemented a simple one as an example (current year) :)

This way other strategies can be easily implemented (and automatically used) later :)

goller commented on rfc3164/machine.go.rl:
```diff
@@ -0,0 +1,204 @@
+timestamp = (datemmm sp datemday sp hhmmss) >mark %set_timestamp @err(err_timestamp);
+
+# (todo) > RFC3164 says something about its maximum length?
+hostname = hostnamerange >mark %set_hostname $err(err_hostname);
```
If I understand https://tools.ietf.org/html/std13 correctly, 255 octets is the max with hostname and domain.

However, RFC 3164 says `The Domain Name MUST NOT be included in the HOSTNAME field`.
I interpret this to mean that the max HOSTNAME is the max label in STD 13: 63 octets.

I think `hostnamerange` with 255 graphs is the right way to go.  Having extra space more safe I think.

leodido replied:
Oh, I see.

Anyway, I would leave this in [1,255]. At least in this PR.

An alternative (for a follow-up PR) would be to provide an option (eg. `WithoutDomainNameIntoHostname`?) to the parser to make it stricter ([1,63] octets).

WDYT?

leodido replied:
In the meantime, I insert a comment for future reference.

goller replied:
Yeah, I could see giving more flexibility with those options in the future.  Good thinkin!

goller commented on rfc3164/machine.go.rl:
```diff
@@ -0,0 +1,204 @@
+hostname = hostnamerange >mark %set_hostname $err(err_hostname);
+
+# Section 4.1.3
+tag = (print - [ :\[]){1,32} >mark %set_tag @err(err_tag); # note(leodido) > alnum{1,32} is too restrictive (no dashes, see https://tools.ietf.org/html/rfc2234#section-2.1)
```
In looking at `docs/rfc3164_tag.dot` the grammar allows

```
!..9
;..Z
\..~
```

The RFC says the following about tags:

```
   The TAG is a string of
   ABNF alphanumeric characters that MUST NOT exceed 32 characters.  Any
   non-alphanumeric character will terminate the TAG field and will be
   assumed to be the starting character of the CONTENT field
```

Should the tag grammar be further restricted?

I also see more information in [5.3](https://tools.ietf.org/html/rfc3164#section-5.3):

```
   Quite often, additional information is included at the beginning of
   the CONTENT field.  The format of "TAG[pid]:" - without the quote
   marks - is common.  The left square bracket is used to terminate the
   TAG field in this case and is then the first character in the CONTENT
   field.
```

leodido replied:
This was my concern too.

In fact, I initially used `alnum{1,32}` as you can see in the comment. Which is more restrictive than `print`.

Looking at various BSD-Syslog messages online, and looking at the interpretation of "ABNF alphanumeric" into [RFC 2234](https://tools.ietf.org/html/rfc2234#section-2.1), I decided to relax the grammar here using `print` (except starter characters for _content_ part).

The tag is conventionally used to store process names, and there are a lot of process names that contain dashes. Should we try using something in the middle? Like only alphanumeric plus dashes plus underscores and not all the print characters?

WDYT Chris? :)

goller replied:
Good point.  I like alphanumeric/dashes/underscores.  I also think period, "."; for example, "howdy.sh"

I'm debating if we should just use what you have.  Ultimately, tag would be nearly any string until a semicolon or [.  

Now that I understand your reasoning, I think you should keep what you have already.  Do you mind adding a comment that "justifies" it?  Something like,
"while RFC3164 assumes only alphanumeric process names, many processes have additional characters."

leodido replied:
Agree!

leodido replied:
Commented the grammar ✅

goller commented on rfc3164/machine.go.rl:
```diff
@@ -0,0 +1,204 @@
+
+fail := (any - [\n\r])* @err{ fgoto main; };
+
+main := pri timestamp sp hostname sp msg;
```
Do you mind adding a comment that this library requires (for now) PRI, TIMESTAMP, and HOSTNAME?

I don't think this PR needs to handle sections [4.3.2](https://tools.ietf.org/html/rfc3164#section-4.3.2) and [4.3.3](https://tools.ietf.org/html/rfc3164#section-4.3.3).

goller replied:
Do you imagine in the future something like this?

```
(pri (timestamp (sp hostname sp)?)?)? msg;
```

leodido replied:
I agree that this PR does not need to handle 4.3.2 (valid PRI + no timestamp/invalid timestamp) and 4.3.3 (no or unidentifiable PRI).

Nevertheless, I think this parser should catch these situations - eg., throwing (non-generic) errors.
So that later, a user implementing a relay, or simply a wrapper around this parser, could detect those errors (for example no PRI, 4.3.3) and behave accordingly.

Clearly, this is easily doable as long I implement best-effort mode for this parser too. In this way when errors occur we can still recover the good parts of the message (useful in 4.3.2 use case for example).

Long story short: no, honestly, I do not imagine having **all** that pieces optional in the future.

goller replied:
Smart.  I like that point.  Use errors.

goller commented on rfc3164/machine.go.rl:
```diff
@@ -0,0 +1,204 @@
+
+content = '[' contentval ']'; # todo(leodido) > support ':' and ' ' too. Also they have to match?
+
+mex = visible+ >mark %set_message;
```
I think this is technically `content`

```
content = visible+ >mark %set_message;
```

and content is technically pid... maybe this?

```
 pidval = !alnum @err(err_contentstart) >mark print* %set_pidval @err(err_pidval);
 pid = '[' pidval ']'; # todo(leodido) > support ':' and ' ' too. Also they have to match?
```

leodido replied:
Chris this is not clear to me. Not getting it, sorry.

Why are you suggesting of renaming "mex" into "content" ?

goller replied:
Yeah, I'm thinking mex into content, but, only because I was hoping to see pid/pidval.  Not that important to change I think.

leodido replied:
Yeah, I just used tag and content as in the RFC3164.
Not that important imho :)

goller commented on rfc3164/machine.go.rl:
```diff
@@ -0,0 +1,204 @@
+}
+
+action err_hostname {
+	m.err = fmt.Errorf(errHostname, m.p)
```
Does it matter to have a mix of tab and spaces?

leodido replied:
Gotta format it manually. Same old story 🤦‍♂ 😆 

Will do it at the end :)

leodido replied:
Done

goller commented on rfc3164/machine_test.go:
```diff
@@ -0,0 +1,68 @@
+		"",
+		nil,
+	},
+	// todo > other test cases pleaaaase
```
We could add these test cases as a follow-on PRs.  What do you think?

leodido replied:
I'd prefer having 10 here covering most common messages. Maybe grabbing them from somewhere lese (we also have an issue here with some examples anyway).

Would you help me doing it? :)

goller replied:
@leodido Yup, I'll grab log messages.  Should I just add commits to this PR for the test cases?

leodido replied:
Maybe we can add a bit of them and then merge it :)

goller commented on rfc3164/parser.go:
```diff
@@ -0,0 +1,45 @@
+	syslog "github.com/influxdata/go-syslog/v2"
+)
+
+// parser represent a RFC3162 parser with mutex capabilities.
```
Probably 3164 :)  I did take a minute and read a bit about [3162](https://tools.ietf.org/html/rfc3162) though

leodido replied:
Tnx

goller reviewed (changes requested):
As usual, great stuff @leodido :)

I read the RFC while reviewing this PR and I spent most of my time checking the grammar.  The dot files really helped!

Regarding code structure changes (interface, base, builder), I think that's all just fine and great!  As you mentioned to me, this will require a new major release version.

Most of my suggestions were based on my interpretation of the RFC.  I think making PRI and HOSTNAME optional could be work in follow-on PRs.  Let me know what you think!

leodido committed:
update(common): refer to RFC 5427 for facilities

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
update: facility test

leodido committed:
fix: RFC 3164 requires a leading space for month day less than 10

leodido committed:
update: RFC3164 grammar to support month day (less than 10) with leading space

leodido committed:
docs: commenting choices made for BSD-syslog parsing

leodido committed:
chore: formatting RFC3164 rnd RFC5424 ragel grammar

Spaces to tabs

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
docs(rfc3164): go examples

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
new(rfc3164): year strategies

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
new(rfc3164): introducing WithYear option for RFC3164 parsers and machines

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
update(rfc3164): update 3164 FSM to support options and add WithYear option to it

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
docs(rfc3164): example using current year strategy

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
docs(rfc3164): best effort for RFC3164 example

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

russorat commented:
Really looking forward to this!!!

goller committed:
test(rfc3164): add example sudo log

Note: I believe this test shows a bug in the parser.
Signed-off-by: Chris Goller <goller@gmail.com>

leodido committed:
update(rfc3164): strict subtraction for tag terminators/content starters

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
fix(rfc3164): msg needs a whitespace as starting character

leodido committed:
fix(rfc3164): add year to Stamp timestamp

leodido commented:
Hey, @goller your test (https://github.com/influxdata/go-syslog/pull/27/commits/db2b9df413ba9a2f76f4d4cdec4b0e07d8eda753) was not passing because of missing whitespace at the start of the MSG part.

If you think that such BSD-syslog messages (no whitespace space between the TAG/CONTENT end - ':' - and the MSG start) are very common we could think to provide an option to parse them as correct ones.

leodido committed:
docs(README): update README to match interfaces, and add RFC3164 (fixes #26)

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
new(rfc3164): other simple test cases

leodido committed:
new(rfc3164): basic strategy to explicitly specify an year for stamp timestamps

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
new(rfc3164): introducing option WithRFC3339

Since there are malformed RFC3164 syslog containing RFC3339 timestamps rathern than STAMP ones, I introduced this option to let user decide whether to parser should ALSO look for them or not.

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
docs(rfc3164): examples RFC3164 containing RFC3339 timestamp matching when using WithRFC3339 option

Note also that STAMP timestamps always match: the `WithRFC3339` option DOES NOT disable the correct behavior (as per RFC 3164).

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
new(rfc3164): option to set the time zone on the Stamp timestamps of the syslog messages

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
docs(rfc3164): examples for WithTimezone option

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido committed:
fix(rfc3164): timezone examples

Signed-off-by: Leonardo Di Donato <leodidonato@gmail.com>

leodido commented:
Friendly ping! @goller 😃

russorat commented:
is this really happening?! I'm so excited! Thank you for the great work here @leodido !

goller reviewed (approved):
LGTM... so many wonderful things :)
